subprojects/core/src/main/java/org/gradle/execution/taskgraph/DefaultTaskExecutionPlan.java

```diff
@@ -484,36 +482,98 @@ public TaskInfo getTaskToExecute() {
                         tasksCancelled = true;
                     }
                 }
-                TaskInfo nextMatching = null;
-                boolean allTasksComplete = true;
-                Iterator<TaskInfo> iterator = executionQueue.iterator();
+
+                final Iterator<TaskInfo> iterator = executionQueue.iterator();
+                final AtomicBoolean selected = new AtomicBoolean();
+                final AtomicBoolean restartIteration = new AtomicBoolean();
+                final Set<String> lockedProjects = Sets.newHashSet();
                 while (iterator.hasNext()) {
-                    TaskInfo taskInfo = iterator.next();
-                    allTasksComplete = allTasksComplete && taskInfo.isComplete();
-                    if (taskInfo.isReady() && taskInfo.allDependenciesComplete() && canRunWithWithCurrentlyExecutedTasks(taskInfo)) {
-                        nextMatching = taskInfo;
-                        iterator.remove();
-                        break;
+                    final TaskInfo taskInfo = iterator.next();
+                    Project project = taskInfo.getTask().getProject();
+                    final String projectPath = ((ProjectInternal) project).getIdentityPath().toString();
```

Comment:
Actually, I think we do.  `taskInfo.getTask()` returns `Task` and `Task.getProject()` returns `Project`.